lib/elixir/test/elixir/code_fragment_test.exs

```diff
@@ -554,6 +570,50 @@ defmodule CodeFragmentTest do
       end
     end
 
+    test "underscored special forms" do
+      assert CF.surround_context("__MODULE__", {1, 1}) == %{
+               context: {:local_or_var, '__MODULE__'},
+               begin: {1, 1},
+               end: {1, 11}
+             }
+
+      assert CF.surround_context("__MODULE__.Foo", {1, 12}) == %{
+               context: {:alias, {:local_or_var, '__MODULE__'}, 'Foo'},
+               begin: {1, 1},
+               end: {1, 15}
+             }
+
+      assert CF.surround_context("%__MODULE__{}", {1, 5}) == %{
+               context: {:struct, {:local_or_var, '__MODULE__'}},
+               begin: {1, 1},
+               end: {1, 12}
+             }
+
+      assert CF.surround_context("%__MODULE__.Foo{}", {1, 13}) == %{
+               context: {:struct, {:local_or_var, '__MODULE__'}, 'Foo'},
```

Comment:
Those changes are looking really good. But I think this should be:

```suggestion
               context: {:struct, {:alias, {:local_or_var, '__MODULE__'}, 'Foo'}},
```

I.e. I think we should change struct to be permanently `{struct, parent}`. :)

Reply:
I was also thinking of that. What to do about cases when `{:struct, ''}` is returned?

Reply:
`{:struct, :none}` perhaps?

Reply:
Actually, I think the most consistent would be simply `:struct`. After all, we do `{:alias, 'Foo'}`, and not `{:alias, :none, 'Foo'}`.

Reply:
Then there are in cursor_context cases such as
```
%Some.
%__MODULE__.
%__MODULE__.Sub.
```
the first one used to return `{:struct, 'Some.'}` - inconsistent with what `Some.`, which returns `{:dot, {:alias, 'Some'}, ''}`

Reply:
Yeah, I think we need to make `{:struct, 'Some.'}` become `{:struct, {:dot, {:alias, 'Some'}, ''}}`, because at that point we indeed don't know if the user will type `Some.macro()` or `Some`. This means the existing `{:struct, charlist}` cases will become at least:

  :struct
  {:struct, {:alias, charlist}}
  {:struct, {:dot, {:alias, charlist}, ''}}

Those are all the literal cases. Then the cases with `__MODULE__` are more complex though.

We probably need a specific section in the docs about structs with several examples.

Reply:
On the other hand, a struct will be an alias at the majority of the time, so should we show functions in the autocompletion? 🤔

Reply:
Check this out, we actually have a bug today:

```
iex(2)> Code.Fragment.cursor_context "%Foo.nam"
{:struct, 'Foo.'}
```

Reply:
Do you think we should return `:none` or return something for example the above?

---

Here is another suggestion. Keep `{:struct, charlist}` for aliases. For everything else, keep `{:struct, inside_dot}`. So "%Foo.nam" should be `{:struct, {:dot, {:alias, 'Foo'}, 'nam'}}`. I think this is what you originally wrote. So no incompatible changes and it makes it opt-in for users if they want to handle the non-alias case.

Similarly, `%__MODULE__` returns `{:struct, {:local_or_var, '__MODULE__'}}`. And `%__MODULE__.Foo` returns `{:struct, {:alias, {:local_or_var, '__MODULE__}, 'Foo'}}`.

WDYT?

Reply:
> On the other hand, a struct will be an alias at the majority of the time, so should we show functions in the autocompletion?

I don't remember how it works currently in elixir_sense. In my opinion we should not overcomplicate here and focus on common patterns. `__MODULE__` is a common one, attributes set to module also (though not so much with structs). I think I haven't seen macros returning alias parts in real code :)

> Here is another suggestion...

I like that, will give it a try